torch/utils/_benchmark/utils/valgrind_wrapper/timer_interface.py

```diff
@@ -0,0 +1,448 @@
+                    with open(error_log, "rt") as f:
+                        error_report = f.read()
+                if not error_report:
+                    error_report = "Unknown error.\n" + valgrind_invocation_output
+
+                raise OSError(f"Failed to collect callgrind profile:\n{error_report}")
+
+            def parse_output(inclusive: bool):
+                annotate_invocation, annotate_invocation_output = run([
+                    "callgrind_annotate",
+                    f"--inclusive={'yes' if inclusive else 'no'}",
+                    callgrind_out
+                ], check=True)
+
+                begin_collecting = False
+                fn_counts = []
+                for l in annotate_invocation_output.splitlines(keepends=False):
+                    if not begin_collecting and re.match(r"Ir\s+file:function", l):
+                        begin_collecting = True
+                        continue
+
+                    count_match = re.match(r"^\s*([0-9,]+)\s+(.+:.+)$", l)
+                    if count_match:
+                        ir_str, file_function = count_match.groups()
+                        ir = int(ir_str.replace(",", ""))
+                        fn_counts.append((ir, file_function))
+                        continue
+
+                    if begin_collecting and re.match(r"-+", l):
+                        continue
+
+                    begin_collecting = False
+
+                return fn_counts
+            return parse_output(inclusive=True), parse_output(inclusive=False)
+        finally:
+            shutil.rmtree(working_dir)
+
+    @staticmethod
+    def _construct_script(stmt: str, setup: str, number: int, num_threads: int, error_log: str, stat_log: str):
+        # The naive template looks something like:
+        #   "for _ in range({number}): {stmt}"
+        # However a loop in Python is surprisingly expensive, and significantly
+        # increases the number of background Python instructions. So instead we
+        # partially unroll the loops, with a block size of 100 chosen to keep
+        # the instruction overhead from `range` low while also not ballooning
+        # the size of the generated file.
+        block_size = 100
+        loop_count = number // block_size
+        remainder = number - block_size * loop_count
+        blocked_stmt = ""
+        if loop_count:
+            unrolled_stmts = textwrap.indent("\n".join([stmt] * block_size), " " * 4)
+            blocked_stmt += f"for _ in range({loop_count}):\n{unrolled_stmts}\n"
+        if remainder:
+            blocked_stmt += "\n".join([stmt] * remainder)
+
+        return textwrap.dedent(r"""
+            import gc
+            import os
+            import subprocess
+            import sys
+            import time
+
+            import torch
+            torch.set_num_threads({num_threads})
+
+            PID = os.getpid()
+
+            def log_failure(msg):
+                with open("{error_log}", "wt") as f:
```

Comment:
If `error_log` has a quote in it this will break out of the string. (Many many years of working on PHP security vulnerabilities makes me very sensitive to this lol). `repr()` should DTRT, just drop the quotes then.

Reply:
On the one hand, this should (???) never come up since it would require tempfile to put a quote in the path. OTOH, this is beautifully elegant and I love it.